### DISCUSSION TITLE
Mark Node As Modified

### DISCUSSION THREAD

**Discussion by maheshrajrp** [eclipse-theia/theia#10864] [Q&A] [answered]
Hi,
I'm trying mark TreeNode as modified and same for editorTabs. Is there any theia implementation which I could refer to ?

Thanks.

**Comment by vince-fugnitto** [comment, marked as answer]
Hi @maheshrajrp, you'll likely want to take a look at [TreeDecorator](https://eclipse-theia.github.io/theia/docs/next/interfaces/core.TreeDecorator-2.html) which gives you the ability to decorate the nodes of a tree.

We have the [ProblemDecorator](https://github.com/eclipse-theia/theia/blob/master/packages%2Fmarkers%2Fsrc%2Fbrowser%2Fproblem%2Fproblem-decorator.ts#L35) as an example you can look at, it currently decorates the resources of the navigator (explorer) tree when the resource in question has problem markers (ex: errors):

![image](https://user-images.githubusercontent.com/40359487/157448242-a08dbd29-d3a0-4d6f-b158-ea3bfcd7e252.png)

Is this similar to what you want to achieve? Modified tabs more closely resemble the [`scm`](https://github.com/eclipse-theia/theia/blob/334b92eed8b1c76f2c062a3891530e631107b7a5/packages/scm/src/browser/decorations/scm-navigator-decorator.ts#L29) decorator which is already implemented.

**Reply by maheshrajrp** [reply]
Hi @vince-fugnitto ,
Thank you for your reply. Yes, those match and I'll refer those for my current implementation. 
Also, I was hoping if you could share the file which I could refer to for this functionality ~ Marking the editor tab as modified. So, that it will ask if the user wishes to save. 
![image](https://user-images.githubusercontent.com/80001447/157604806-a457df38-5335-4245-a979-503eaf1d907c.png)
 

Thank you.

**Reply by vince-fugnitto** [reply]
You can take a look at what is known as [TabBarDecorator](https://eclipse-theia.github.io/theia/docs/next/interfaces/core.TabBarDecorator.html) with [ProblemTabBarDecorator](https://github.com/eclipse-theia/theia/blob/e62fd33ce46055d1633d781c08bf5852535f53c5/packages/markers/src/browser/problem/problem-tabbar-decorator.ts#L29) being an example.

The example you mentioned is pretty internal to the framework, it marks any dirty tab with the dirty marker, you'll likely want to go the decorator route for your use-case. If you're still interested it is done at [saveable.ts](https://github.com/eclipse-theia/theia/blob/e62fd33ce46055d1633d781c08bf5852535f53c5/packages/core/src/browser/saveable.ts#L269-L279).

**Reply by maheshrajrp** [reply]
Sorry for the late reply. This really helps. Thank you.
